File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/uncomment/direct/Java/Python/s685734377_commented.py
def main():
    import sys
    n = int(sys.stdin.readline())
    a = [sys.stdin.readline().strip() for _ in range(n)]
    P = 37
    Q = 47
    powP = [1] * (n + 1)
    powQ = [1] * (n + 1)
    for i in range(1, n + 1):
        powP[i] = powP[i - 1] * P
        powQ[i] = powQ[i - 1] * Q

    hashRow = [[0] * n for _ in range(n)]
    hashCol = [[0] * n for _ in range(n)]

    for i in range(n):
        for j in range(n):
            for k in range(n):
                exponent_p = (j - k + n) % n
                term = ord(a[i][j]) * powP[exponent_p]
                hashRow[i][k] += term

                exponent_p_col = (i - k + n) % n
                term_col = ord(a[i][j]) * powP[exponent_p_col]
                hashCol[j][k] += term_col

    ans = 0
    for A in range(n):
        for B in range(n):
            ok = True
            for i in range(n):
                row_idx = (i + A) % n
                col_idx = (i + B) % n
                if hashRow[row_idx][B] != hashCol[col_idx][A]:
                    ok = False
                    break
            if ok:
                ans += 1
    print(ans)

if __name__ == "__main__":
    main()
# End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/uncomment/direct/Java/Go/s685734377_commented.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

type FastScanner struct {
	scanner *bufio.Scanner
	tokens  []string
	current int
}

func NewFastScanner(r io.Reader) *FastScanner {
	fs := &FastScanner{
		scanner: bufio.NewScanner(r),
		tokens:  make([]string, 0),
	}
	fs.scanner.Split(bufio.ScanWords)
	return fs
}

func (fs *FastScanner) Next() string {
	if fs.current < len(fs.tokens) {
		token := fs.tokens[fs.current]
		fs.current++
		return token
	}
	if !fs.scanner.Scan() {
		return ""
	}
	fs.tokens = strings.Fields(fs.scanner.Text())
	fs.current = 1
	return fs.tokens[0]
}

func (fs *FastScanner) NextInt() int {
	s := fs.Next()
	n, _ := strconv.Atoi(s)
	return n
}

type TaskB struct{}

func NewTaskB() *TaskB {
	return &TaskB{}
}

func (t *TaskB) solve(in *FastScanner, out *bufio.Writer) {
	n := in.NextInt()
	grid := make([][]byte, n)
	for i := 0; i < n; i++ {
		s := in.Next()
		grid[i] = []byte(s)
	}

	const P = 37
	const Q = 47
	powP := make([]int64, n+1)
	powQ := make([]int64, n+1)
	powP[0] = 1
	powQ[0] = 1
	for i := 1; i <= n; i++ {
		powP[i] = powP[i-1] * int64(P)
		powQ[i] = powQ[i-1] * int64(Q)
	}

	hashRow := make([][]int64, n)
	hashCol := make([][]int64, n)
	for i := range hashRow {
		hashRow[i] = make([]int64, n)
		hashCol[i] = make([]int64, n)
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			for k := : 0; k < n; k++ {
				deltaRow := (j - k + n) % n
				termRow := int64(grid[i][j]) * powP[deltaRow]
				hashRow[i][k] += termRow

				deltaCol := (i - k + n) % n
				termCol := int64(grid[i][j]) * powP[deltaCol]
				hashCol[j][k] += termCol
			}
		}
	}

	ans := 0
	for A := 0; A < n; A++ {
		for B := 0; B < n; B++ {
			ok := true
			for i := 0; i < n; i++ {
				rowIdx := (i + A) % n
				colIdx := (i + B) % n
				if hashRow[rowIdx][B] != hashCol[colIdx][A] {
					ok = false
					break
				}
			}
			if ok {
				ans++
			}
		}
	}
	fmt.Fprintln(out, ans)
	out.Flush()
}

func main() {
	scanner := NewFastScanner(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	taskB := NewTaskB()
	taskB.solve(scanner, out)
	out.Flush()
}

// End of Code
